### ISSUE TITLE
boost 1.70.0: undefined symbol error when importing espressomd

### ISSUE THREAD

**Issue by xibeisiber** [espressomd/espresso#3055]
Hi,
I am installing espresso on my computer. There is no error message in the progress of "cmake" and "make". But when I import the espressomd, it returns:

File "<stdin>", line 1, in <module>
  File "/remote/pi309a/jia/software/Espresso/src/python/espressomd/__init__.py", line 23, in <module>
    import espressomd._init
ImportError: /remote/pi309a/jia/software/Espresso/src/core/cluster_analysis/libcluster_analysis.so.4: undefined symbol: _ZTIN5boost7archive6detail14basic_iarchiveE

The output of cmake is:
[cmake.txt](https://github.com/espressomd/espresso/files/3468305/cmake.txt)
The output of make is:
[make.txt](https://github.com/espressomd/espresso/files/3468306/make.txt)
I don't know what is going on. Several months ago I installed it, all is fine..
My problem is similar to the case in another post, but his way to solve it does not apply to me..
Thanks very much for any help!
Best Wishes!
Jia

**Comment by jngrad** [comment]
Hi,
We'll need your operating system (Ubuntu 18.04, CentOS 7, ...) and the version of espresso (4.0.2, devel, ...) to reproduce your issue.

**Comment by xibeisiber** [comment]
> Hi,
> We'll need your operating system (Ubuntu 18.04, CentOS 7, ...) and the version of espresso (4.0.2, devel, ...) to reproduce your issue.

Hi, jngrad, 
thanks. the operating system is Debian GNU/Linux 9.9 (stretch).
I tried espresso 4.0.2 and 4.0.1. Both versions have the same problem.

**Comment by jngrad** [comment]
I just compiled espresso 4.0.2 for Python2 in Debian 9 with CMake 3.15.1 and boost 1.70.0 and got the same error message, but instead of `typeinfo for boost::archive::detail::basic_iarchive`, got a different undefined symbol: `boost::filesystem::detail::copy_file(boost::filesystem::path const&, boost::filesystem::path const&, boost::filesystem::detail::copy_option, boost::system::error_code*)`. I had specifically compiled boost with `system` and `filesystem`. There was also no warning during compilation of es. Compiling 4.0.2 for boost 1.69 went just fine.

For other core es developers: the error can be reproduced in docker with image [debian-python2:9](https://github.com/espressomd/docker/blob/4.0/docker/debian/Dockerfile-9), removing line 11 and adding the following:
```bash
# Boost
RUN cd /tmp \
 && mkdir boost \
 && cd boost \
 && curl -sL https://dl.bintray.com/boostorg/release/1.70.0/source/boost_1_70_0.tar.bz2 | tar xj \
 && cd boost_1_70_0 \
 && echo 'using mpi ;' > tools/build/src/user-config.jam \
 && ./bootstrap.sh --with-libraries=filesystem,mpi,serialization,test,system \
 && ./b2 -j $(nproc) cxxflags=-fPIC install --prefix=/opt/boost \
 && cd \
 && rm -r /tmp/boost

ENV BOOST_ROOT=/opt/boost LD_LIBRARY_PATH="/opt/boost/lib:${LD_LIBRARY_PATH}"

# CMake
RUN cd /usr/local \
&& curl -sL https://cmake.org/files/v3.15/cmake-3.15.1-Linux-x86_64.tar.gz | tar --strip-components=1 -xz \
&& rm -r /usr/local/man
```
the `-fPIC` flag is required to avoid this error message while compiling espresso:
```
/usr/bin/ld: /opt/boost/lib/libboost_mpi.a(communicator.o): relocation R_X86_64_PC32 against symbol `ompi_mpi_comm_world' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: final link failed: Bad value
```

**Comment by jngrad** [comment]
I also can't run unit tests with 1.70.0 because of another undefined symbol: ```undefined reference to `boost::unit_test::unit_test_main(bool (*)(), int, char**)'```

**Comment by KaiSzuttor** [comment]
https://stackoverflow.com/questions/35007134/c-boost-undefined-reference-to-boostfilesystemdetailcopy-file

**Comment by jngrad** [comment]
actually, we already use that workaround:
https://github.com/espressomd/espresso/blob/5322996769ec0ff72ba82cce276ef723fc2edcd4/src/core/io/writer/h5md/h5md_core.hpp#L35-L37
defining it globally with `cmake .. -DCMAKE_CXX_FLAGS=-DBOOST_NO_CXX11_SCOPED_ENUMS` didn't help

**Comment by xibeisiber** [comment]
Several months ago I compiled espresso 4.0.1 successfully, the CMakeCache.txt file is as follows:_----------
//The directory containing a CMake configuration file for Boost.
Boost_DIR:PATH=Boost_DIR-NOTFOUND

//Boost filesystem library (debug)
Boost_FILESYSTEM_LIBRARY_DEBUG:FILEPATH=/usr/lib/x86_64-linux-gnu/libboost_filesystem.so

//Boost filesystem library (release)
Boost_FILESYSTEM_LIBRARY_RELEASE:FILEPATH=/usr/lib/x86_64-linux-gnu/libboost_filesystem.so

//Path to a file.
Boost_INCLUDE_DIR:PATH=/usr/include

//Boost library directory DEBUG
Boost_LIBRARY_DIR_DEBUG:PATH=/usr/lib/x86_64-linux-gnu

//Boost library directory RELEASE
Boost_LIBRARY_DIR_RELEASE:PATH=/usr/lib/x86_64-linux-gnu

//Boost mpi library (debug)
Boost_MPI_LIBRARY_DEBUG:FILEPATH=/usr/lib/x86_64-linux-gnu/libboost_mpi.so

//Boost mpi library (release)
Boost_MPI_LIBRARY_RELEASE:FILEPATH=/usr/lib/x86_64-linux-gnu/libboost_mpi.so

//Boost serialization library (debug)
Boost_SERIALIZATION_LIBRARY_DEBUG:FILEPATH=/usr/lib/x86_64-linux-gnu/libboost_serialization.so

//Boost serialization library (release)
Boost_SERIALIZATION_LIBRARY_RELEASE:FILEPATH=/usr/lib/x86_64-linux-gnu/libboost_serialization.so

//Boost system library (debug)
Boost_SYSTEM_LIBRARY_DEBUG:FILEPATH=/usr/lib/x86_64-linux-gnu/libboost_system.so

//Boost system library (release)
Boost_SYSTEM_LIBRARY_RELEASE:FILEPATH=/usr/lib/x86_64-linux-gnu/libboost_system.so

//Boost unit_test_framework library (debug)
Boost_UNIT_TEST_FRAMEWORK_LIBRARY_DEBUG:FILEPATH=/usr/lib/x86_64-linux-gnu/libboost_unit_test_framework.so

//Boost unit_test_framework library (release)
Boost_UNIT_TEST_FRAMEWORK_LIBRARY_RELEASE:FILEPATH=/usr/lib/x86_64-linux-gnu/libboost_unit_test_framework.so

This time this part is replaced by 
-------
//No help, variable specified on the command line.
BOOST_ROOT:STRING=/usr/include

//The directory containing a CMake configuration file for Boost.
Boost_DIR:PATH=/opt/boost_1_70_0/lib/cmake/Boost-1.70.0

//No help, variable specified on the command line.
Boost_NO_SYSTEM_PATHS:UNINITIALIZED=TRUE
......
I'm not sure why they are different, maybe the IT administer in my institute upgraded my computer...

**Comment by jngrad** [comment]
Debian 9 is shipped with [libboost v1.62](https://packages.debian.org/stretch/libboost-dev) (stored in `/usr/lib/x86_64-linux-gnu/`), maybe your institute decided to move to libboost v1.70, which means manually compiling the source code (by default it gets installed in a different directory: `/opt/boost_1_70_0/`). If you still have the 1.62 files (check with `ls /usr/lib/x86_64-linux-gnu/libboost*`), you may be able to use them instead of the newer 1.70 by tweaking the boost search paths in CMake.

**Comment by xibeisiber** [comment]
Thanks very much, jngrad.
Problem solved now by forcing cmake to use boost 1.62.0.
